src/gpuarray_buffer_cuda.c

```diff
@@ -484,17 +482,15 @@ static int cuda_write(gpudata *dst, size_t dstoff, const void *src,
 
     cuda_enter(ctx);
 
-    ctx->err = cuEventSynchronize(dst->ev);
-    if (ctx->err != CUDA_SUCCESS) {
-      cuda_exit(ctx);
-      return GA_IMPL_ERROR;
-    }
+    cuda_wait(dst, CUDA_WAIT_WRITE);
 
-    ctx->err = cuMemcpyHtoD(dst->ptr + dstoff, src, sz);
+    ctx->err = cuMemcpyHtoDAsync(dst->ptr + dstoff, src, sz, ctx->s);
```

Comment:
Basically it does the copy to a staging buffer and returns.  Then the actual copy to the GPU will be done later so we can go on scheduling some other things and not wait for the copy to finish.

However the documentation states that the copy to the staging buffer happens in a worker thread, so maybe there is some time after it returns where it isn't actually safe to touch the CPU memory which is why I don't want to merge this until we are sure.